fastqpreprocessing/src/metricgatherer.cpp

```diff
@@ -0,0 +1,443 @@
+     this->reads_mapped_uniquely += 1;
+   else
+     this->reads_mapped_multiple += 1;  // without multi-mapping, this number is zero!
+
+   this->duplicate_reads += std::stoi(std::string(record[offset + 11]));
+
+   // cigar N field (3) indicates a read is spliced if the value is non-zero
+   this->spliced_reads += std::stoi(std::string(record[offset + 10]));
+
+   prev_tag = current_tag;
+}
+
+
+//  Calculate metrics that require information from all molecules of an entity
+//  ``finalize()`` replaces attributes in-place that were initialized by the constructor as
+//  ``None`` with a value calculated across all molecule data that has been aggregated.
+
+void Metrics::finalize(std::set<std::string> &mitochondrial_genes) {
+   this->molecule_barcode_fraction_bases_above_30_mean = this->_molecule_barcode_fraction_bases_above_30.getMean();
+
+   this->molecule_barcode_fraction_bases_above_30_variance = this->_molecule_barcode_fraction_bases_above_30.calculate_variance();
+
+   this->genomic_reads_fraction_bases_quality_above_30_mean = this->_genomic_reads_fraction_bases_quality_above_30.getMean();
+
+   this->genomic_reads_fraction_bases_quality_above_30_variance = this->_genomic_reads_fraction_bases_quality_above_30.calculate_variance();
+
+   this->genomic_read_quality_mean = this->_genomic_read_quality.getMean();
+
+   this->genomic_read_quality_variance = this->_genomic_read_quality.calculate_variance();
+
+   this->n_molecules = this->_molecule_histogram.size();
+
+   this->n_fragments = this->_fragment_histogram.size();
+
+   try {
+     this->reads_per_molecule = this->n_reads / this->n_molecules;
+   } catch (runtime_error& e) {
+     this->reads_per_molecule = -1;   // float("nan")
+   }
+  
+   try {
+     this->reads_per_fragment  = this->n_reads / this->n_fragments;
+   } catch (runtime_error& e) {
+     this->reads_per_fragment = -1; //float("nan")
+   }
+
+   try {
+     this->fragments_per_molecule = this->n_fragments / this->n_molecules;
+   } catch (runtime_error& e) {
+     this->fragments_per_molecule = -1; // float("nan")
+   }
+
+  this->fragments_with_single_read_evidence = 0; 
+  for (auto local_cit = this->_fragment_histogram.cbegin(); 
+       local_cit!=this->_fragment_histogram.cend(); local_cit++)
+     if (local_cit->second==1)
+        this->fragments_with_single_read_evidence++; 
+
+
+  this->molecules_with_single_read_evidence = 0;
+  for (auto local_cit = this->_molecule_histogram.cbegin(); 
+       local_cit!=this->_molecule_histogram.cend(); local_cit++)
+     if (local_cit->second==1)
+        this->molecules_with_single_read_evidence++; 
+}
+
+////////////////  CellMetrics ////////////////////////
+std::string CellMetrics::getHeader() {
+  std::string s="";
+
+  for (int i=0; i<24; i++) {
+    s += std::string(",") + headers[i];
+  }
+
+  for (int i=0; i<11; i++) {
+    s += std::string(",") + specific_headers[i];
+  }
+
+  return s;
+}
+
+// Parses a record to extract gene-specific information
```

Comment:
Use multi line comment instead of many single line comments

Reply:
done.